packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart

```diff
@@ -19,26 +19,72 @@ import '../../../../../shared/memory/classes.dart';
 import '../../../../../shared/memory/heap_graph_loader.dart';
 import '../../../../../shared/memory/retaining_path.dart';
 import '../../../../../shared/memory/simple_items.dart';
+import '../../../../../shared/primitives/serialization.dart';
 import '../../../shared/heap/class_filter.dart';
 import '../../../shared/primitives/memory_utils.dart';
 import '../data/classes_diff.dart';
 import '../data/csv.dart';
 import '../data/heap_diff_data.dart';
 import '../data/heap_diff_store.dart';
 import 'class_data.dart';
-import 'item_controller.dart';
+import 'snapshot_item.dart';
+
+class _Json {
+  static const snapshots = 'snapshots';
+  static const diffWith = 'diffWith';
+}
 
 class DiffPaneController extends DisposableController {
-  DiffPaneController({required this.loader});
+  DiffPaneController({
+    required this.loader,
+    List<SnapshotDataItem>? snapshots,
+  }) {
+    if (snapshots != null) {
+      core._snapshots.addAll(snapshots);
+    }
+    derived._updateValues();
+  }
 
   factory DiffPaneController.fromJson(Map<String, dynamic> json) {
-    // TODO(polina-c): implement, https://github.com/flutter/devtools/issues/6972
-    return DiffPaneController(loader: null);
+    final snapshots = (json[_Json.snapshots] as List)
+        .map((e) => deserialize(e, SnapshotDataItem.fromJson))
+        .toList();
+
+    final diffWith = json[_Json.diffWith] as List<int?>;
+
+    assert(snapshots.length == diffWith.length);
+
+    for (var i = 0; i < snapshots.length; i++) {
+      final diffIndex = diffWith[i];
+      if (diffIndex != null) {
+        snapshots[i].diffWith.value = snapshots[diffIndex];
+      }
+    }
+
+    return DiffPaneController(
+      loader: null,
+      snapshots: snapshots,
+    );
   }
 
   Map<String, dynamic> toJson() {
-    // TODO(polina-c): implement, https://github.com/flutter/devtools/issues/6972
-    return {};
+    final snapshots = core.snapshots.value
+        .whereType<SnapshotDataItem>()
+        .where((s) => s.heap != null)
+        .toList();
+
+    final snapshotToIndex =
+        snapshots.asMap().map((index, item) => MapEntry(item, index));
+
+    final diffsWithSnapshots = snapshots.map((item) {
```

Comment:
nit: now that I better understand what this var is, how about "diffWithIndices" and then below r/diffWith/diffWithIndex